hack/tpm2/main.go

```diff
@@ -0,0 +1,809 @@
+package main
+
+// TODO: maybe some of this is useful: https://ericchiang.github.io/post/tpm-keys/
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha1"
+	"crypto/sha256"
+	"crypto/sha512"
+	"crypto/x509"
+	_ "embed"
+	"encoding/base64"
+	"encoding/binary"
+	"encoding/hex"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"log"
+
+	"github.com/google/go-tpm/tpm2"
+	"github.com/google/go-tpm/tpm2/transport"
+	"github.com/google/go-tpm/tpm2/transport/simulator"
+	"github.com/google/go-tpm/tpmutil"
+)
+
+var (
+	//go:embed testdata/tpm2-pcr-public.pem
+	pcrSigningPubKey []byte
+	//go:embed testdata/pcr-data.json
+	pcrDataJSON []byte
+)
+
+type LuksHeader struct {
+	Type            string   `json:"type"`
+	Keyslots        []string `json:"keyslots"`
+	TPM2PrivateBlob []byte   `json:"tpm2_private_blob"`
+	TPM2PublicBlob  []byte   `json:"tpm2_public_blob"`
+	TPM2PCRS        []int    `json:"tpm2_pcrs"`
+	TPM2Alg         string   `json:"tpm2_alg"`
+	TPM2PolicyHash  []byte   `json:"tpm2_policy_hash"`
+	TPM2PublicKey   string   `json:"tpm2_public_key"`
+	TPM2SRK         []byte   `json:"tpm2_srk"`
+}
+
+type PCRData struct {
+	SHA1   []bankData `json:"sha1,omitempty"`
+	SHA256 []bankData `json:"sha256,omitempty"`
+	SHA384 []bankData `json:"sha384,omitempty"`
+	SHA512 []bankData `json:"sha512,omitempty"`
+}
+
+type bankData struct {
+	// list of PCR banks
+	PCRS []int `json:"pcrs"`
+	// Public key of the TPM
+	PKFP string `json:"pkfp"`
+	// Policy digest
+	POL string `json:"pol"`
+	// Signature of the policy digest in base64
+	SIG string `json:"sig"`
+}
+
+const (
+	// The SRK handle is defined in the Provisioning Guidance document (see above) in the table "Reserved Handles
+	// for TPM Provisioning Fundamental Elements". The SRK is useful because it is "shared", meaning it has no
+	// authValue nor authPolicy set, and thus may be used by anyone on the system to generate derived keys or
+	// seal secrets. This is useful if the TPM has an auth (password) set for the 'owner hierarchy', which would
+	// prevent users from generating primary transient keys, unless they knew the owner hierarchy auth. See
+	// the Provisioning Guidance document for more details.
+	// https://trustedcomputinggroup.org/resource/tcg-tpm-v2-0-provisioning-guidance
+	// https://trustedcomputinggroup.org/resource/http-trustedcomputinggroup-org-wp-content-uploads-tcg-ek-credential-profile
+	TPM2_SRK_HANDLE = tpmutil.Handle(0x81000001)
```

Comment:
so this is not used in here, since we generate an ECC primary key everytime with the same seed, but as the comment says, if the tpm has auth set for owner hierarchy we might hit a snag, needs discussion